src/serde.rs

```diff
@@ -0,0 +1,153 @@
+use std::{fmt, marker::PhantomData};
+
+use serde::{
+    de::{self, MapAccess, Visitor},
+    ser::{Serialize, SerializeStruct},
+    Deserialize, Deserializer,
+};
+
+use crate::{NodeId, NodeMut, NodeRef, Tree};
+
+#[derive(Debug)]
+struct SerNode<'a, T> {
+    value: &'a T,
+    children: Vec<SerNode<'a, T>>,
+}
+
+impl<'a, T> From<NodeRef<'a, T>> for SerNode<'a, T> {
+    fn from(node: NodeRef<'a, T>) -> Self {
+        let value: &T = node.value();
+        let children: Vec<SerNode<'a, T>> = node.children().map(SerNode::<'a, T>::from).collect();
+        SerNode { value, children }
```

Comment:
Are the type ascriptions necessary? I would expect the function to constrain the types sufficiently, i.e.

```suggestion
        let value = node.value();
        let children = node.children().map(SerNode::from).collect();
        Self { value, children }
```

Of course, please keep them if you added them explicitly to highlight the types for human readers instead of to appease the compiler.